### PR TITLE
feature: add sourcehut "forge" type

### PR DESCRIPTION
**Pull request by rgrinberg**

### CODE REVIEW

File: src/dune_engine/package.ml
Comment (emillon) [inline src/dune_engine/package.ml:323]: The current implementation is fine, but I wonder if this would be even simpler to inline the intermediate abstractions and do a "plain" pattern matching. There will be some duplication, but the various hosts we support are already slightly different.

Comment (rgrinberg) [reply src/dune_engine/package.ml:323]: There's a way to factor this stuff out better most certainly. I did not try very hard I admit.

**Comment by rgrinberg** [comment]
ping @emillon needs an approval

**Review by emillon — APPROVED** [review]

**Comment by mro** [comment]
@rgrinberg do I read the code correct, the source stanza is limited to hardcoded platforms? How would I enter an arbitrary url into an opam file - by dune-project or would I rather just write it by hand?

**Comment by rgrinberg** [comment]
> How would I enter an arbitrary url into an opam file - by dune-project or would I rather just write it by hand?

I think you need the set all the fields individually. Url, homepage, etc. The source stanza is just there to make setting all those fields at once possible.

**Comment by mro** [comment]
https://codeberg.org/seppo/seppo/src/commit/85a9773afbe50360a0a6c17bdb8638cad8446fd3/%23Seppo.opam.template

**Comment by emillon** [comment]
@mro you don't have to use a template for this, the following should work:

```
(homepage Seppo.Social)
(bug_reports Seppo.Social/issues)
(source (uri Seppo.Social/sourcecode))
```

**Comment by emillon** [comment]
Tags are supported, they'll be sourced from `(tags)` as you'd expect.

**Comment by mro** [comment]
failed for me:

```
$ dune build
File "dune-project", line 17, characters 1-5:
17 | (tags Social Web "#Fediverse" ActivityPub CGI RFC3875 RFC4287 RFC5005 RFC7033)
      ^^^^
Error: Unknown field tags
```

**Comment by emillon** [comment]
I think it needs to be in the `(package)` itself.

**Comment by mro** [comment]
it appears a bit inconsequent that entries end up next to each other but one can be specified outside `(package)` while the other has to be within.

**Comment by emillon** [comment]
Yeah I agree it's not ideal. The reasoning is that things like source, author, maintainer are global to the project but tags are specific to each package. But I think that adding a global version of `(tags)` is a reasonable feature request too.

**Pull request merged**
